### ISSUE TITLE
Reference Tables: Foreign Keys

### ISSUE THREAD

**Issue by onderkalaci** [citusdata/citus#968]
Reference tables would have a single shard with replication of (# worker nodes). However, with #698 we considered to not allow foreign keys with replication factor > 1. In that sense, supporting foreign keys for reference tables become tricky. We didn't add this issue to MVP for 6.1.

The main problem with that is a concurrent arbitrary multi-shard modification could run in an order such that the reference table's CASCADE operations are blocked on them and vice versa. See the example below:

1. User runs a DELETE/UPDATE on the `reference_table`
2. The first placement of the `reference_table` gets the DELETE/UPDATE and it cascades the modification to the `shard_1` of  `colocated_table`
3. Then, user runs a concurrent master_modify_multiple_shards (or other type of multi-shard modification) on `colocated_table` on a different connection.
4. The 3rd operation is firstly executed on `shard_2` of `colocated_table`.
5. After that when `reference_table`'s second replica gets the UPDATE/DELETE, it tries to cascade to the `shard_2` of `colocated_table`. However, since the 4th operation holds lock on `shard_2` of `colocated_table`, this operation is blocked.
6. The 3rd operation continues to run on `shard_1` of `colocated_table`. However, since the 2nd operation holds lock on  `shard_1` of `colocated_table`, this operation is blocked.
7. Since both 5 and 6 are blocked, we end-up with a deadlock.

**Comment by onderkalaci** [comment]
Note that we could allow foreign keys among reference tables. The reason is that they'd be single sharded and the above case does not apply for a single shard distributed table.

I'll also double check this with @marcocitus 

This is also semi-related to #962.

**Comment by marcocitus** [comment]
> Note that we could allow foreign keys among reference tables. The reason is that they'd be single sharded and the above case does not apply for a single shard distributed table.

The replication factor issues would still apply, e.g. #861 (also need to take appropriate shard resource locks for cascading delete/update).

**Comment by sumedhpathak** [comment]
Noting example from a contact from our docs repo:

I ran into a related issue with the next table:

```sql
CREATE TABLE my_table (
	id SERIAL NOT NULL,
	name VARCHAR,
	PRIMARY KEY (id),
	UNIQUE (name, id)
);
SELECT create_reference_table('my_table');

CREATE TABLE my_second_table (
	id SERIAL NOT NULL,
	name VARCHAR,
    my_table_id SERIAL NOT NULL,
	PRIMARY KEY (id, my_table_id),
	UNIQUE (name, my_table_id, id),
	FOREIGN KEY(my_table_id) REFERENCES my_table (id) ON DELETE CASCADE
);
SELECT create_reference_table('my_second_table');
```

```
ERROR:  cannot create foreign key constraint
DETAIL:  Partition column must exist both referencing and referenced side of the foreign constraint statement and it must be in the same ordinal in both sides.
```

**Can one reference table _reference_ another?**

When I instead try to distribute `my_second_table`

```sql
SELECT create_distributed_table('my_second_table', 'id');
```

I get the following error:

```
ERROR:  cannot create foreign key constraint
DETAIL:  Foreign key constraint can only be created on co-located tables.
```

**Can a distributed table _reference_ a reference table?**

**Comment by onderkalaci** [comment]
Moving some notes from #969 
* [ ] Foreign Keys
	* [ ] Simply do not allow, given that we decided not to include foreign keys for reference tables in 6.1

**Comment by onderkalaci** [comment]
I was experimenting with foreign keys on reference tables and found a potential deadlock. 

Though the following example currently is NOT supported and NOT planned to be supported in 6.1, we should keep in mind while implementing foreign keys for reference tables in the future.


```SQL

SET citus.shard_replication_factor TO 1;
SET citus.shard_count TO 4;


CREATE TABLE main_1 (x int, y int);
SELECT create_distributed_table('main_1', 'x');
CREATE TABLE reference_1 (a int unique, b int);
SELECT create_reference_table('reference_1');
ALTER TABLE main_1 ADD CONSTRAINT main_1_cs FOREIGN KEY (x) REFERENCES reference_1 (a);
 truncate reference_1 CASCADE;


==> /tmp/logfile_9701 <==
LOG:  duration: 0.458 ms  statement: BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED

==> /tmp/logfile_9700 <==
LOG:  duration: 0.291 ms  statement: BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED

==> /tmp/logfile_9701 <==
LOG:  duration: 4.481 ms  statement: TRUNCATE TABLE public.reference_1_102012 CASCADE

==> /tmp/logfile_9700 <==
LOG:  duration: 5.283 ms  statement: TRUNCATE TABLE public.reference_1_102012 CASCADE
LOG:  duration: 0.377 ms  statement: BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED

==> /tmp/logfile_9701 <==
LOG:  duration: 0.312 ms  statement: BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED

==> /tmp/logfile_9700 <==
LOG:  duration: 0.401 ms  statement: BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED

==> /tmp/logfile_9701 <==
LOG:  duration: 2.745 ms  statement: TRUNCATE TABLE public.main_1_102009 CASCADE

==> /tmp/logfile_9700 <==
LOG:  duration: 5.195 ms  statement: TRUNCATE TABLE public.main_1_102011 CASCADE

```

**Comment by ozgune** [comment]
@byucesoy / @onderkalaci -- This recently came up in a customer conversation.

If we don't support foreign keys on reference tables, could we remove the 6.1 tag from this issue and re-open it?

Is the fix for supporting this fairly involved?

**Comment by byucesoy** [comment]
@ozgune We had a small discussion about this with @onderkalaci and @metdos, it seems that this is not an easy fix.

We do not allow foreign keys on reference tables because replication factor of reference tables are greater than 1. This causes some deadlock problems (#861 and #810).

#810 might be less problematic because reference tables will not be marked as inactive but it still needs some thinking.

#861 is more complicated, but it seems that we can prevent it by taking locks in node order not placement order.

And finally there are CASCADE related issues which are captured in the comments in this issue. I think these are the most challenging ones to handle. PostgreSQL processes CASCADE operations and their order is not deterministic. There is no easy way to enforce an order as far as I know. One option could be preventing processing CASCADE operations by PostgreSQL and taking responsibility of processing CASCADE to Citus, which is not an easy task and we need to spend some time on it to understand whether this will work or not.

Maybe @marcocitus can suggest easier way to overcome these problems?

**Comment by marcocitus** [comment]
> Maybe @marcocitus can suggest easier way to overcome these problems?

I think implementing #861 and dropping CASCADE support for reference tables should suffice.

**Comment by byucesoy** [comment]
> I think implementing #861 and dropping CASCADE support for reference tables should suffice.

After implementing #861, we only need 2-3 lines of code change to drop CASCADE support.

Related question: Do we have existing users who use CASCADE?

**Comment by ozgune** [comment]
@byucesoy -- Got it. I'm reopening this issue because we currently don't support foreign keys on reference tables and we don't have an issue to track it. This came up in a few customer conversations and we manually removed the foreign key constraints in those cases.

**Comment by ozgune** [comment]
+2 This issue came up again in two recent customer conversations. @saicitus -- could you capture the most recent question in our product sheet?

I'm reading this issue's (#968) history, and I had two questions.

* @byucesoy -- I'm reading through your CASCADE support for reference tables note. What does this mean? Could you give an example?
* @onderkalaci / @marcocitus -- How long would the fix for #861 take? Could we rely on the upcoming deadlock detection changes as a workaround?

**Comment by onderkalaci** [comment]
Citus already supports foreign key constraints among co-located hash distributed tables as shown in the related regression test [file](https://github.com/citusdata/citus/blob/master/src/test/regress/expected/multi_foreign_key.out).

The missing piece that we think would be very useful is to allow foreign keys from hash distributed tables to reference tables. As an example:

```SQL
CREATE TABLE referenced_table(id int UNIQUE, PRIMARY KEY(id));
SELECT create_reference_table('referenced_table');

-- test foreign constraint creation with not supported parameters on the distribution key
CREATE TABLE referencing_table(id int, ref_id int, FOREIGN KEY(ref_id) REFERENCES referenced_table(id) ON DELETE SET NULL);
SELECT create_distributed_table('referencing_table', 'ref_id');
```

The main obstacle to implement this change is the potential distributed deadlocks or reading data over a wrong connection (e.g., wrong results) that might happen due to the parallel execution of the commands.

Please note that most of the issues ariase inside a transaction block that touches both reference tables and multiple shards of distributed tables that refer to the reference tables. So, for example I cannot see any complications related to multi-tenant use case where the same transaction modifies reference tables and distributed tables that touch a single shard. Or, any query that is not executed inside a transaction block would be affect by this.

I could probably bucket the transactional issues as follows:

### 1. A modification to a reference table is followed by a multi-shard modify or DDL on a distributed table that has foreign key (or transitive foreign key) to the reference table
```SQL
BEGIN;
UPDATE referenced_table SET id = 15 WHERE id = 2;
UPDATE referencing_table SET ref_id = 3 WHERE ref_id = 2;
ERROR:  canceling the transaction since it was involved in a distributed deadlock
```
 
### 2. A modification to a reference table is followed by a multi-shard real-time `SELECT`  on a distributed table that has foreign key (or transitive foreign key) to the reference table (e.g., read-your-write-consistency)

``` SQL
-- ON UPDATE CASCADE relation exists
INSERT INTO referenced_table VALUES (2);
INSERT INTO referencing_table SELECT i, 2 FROM generate_series(0, 1000) i;

BEGIN;
UPDATE referenced_table SET id = 15 WHERE id = 2;
SELECT count(*) FROM referencing_table WHERE ref_id = 2;
-- the correct result is 0
 count 
-------
   912
(1 row)
```

### 3. A modification to a reference table is followed by a COPY (or coordinator `INSERT ... SELECT` or loading the existing data in local table -- currently errors out but might just work after we support `SELECT FOR UPDATE`)
```SQL
UPDATE referenced_table SET id = 15 WHERE id = 2;
COPY referencing_table FROM '/tmp/data.csv' WITH CSV;
ERROR:  canceling the transaction since it was involved in a distributed deadlock
```

### 4. A modification-DDL-COPY to a distributed table that has foreign key (or transitive foreign key) to the reference table is followed by a modification to a reference table
```SQL
UPDATE referencing_table SET ref_id = 3 WHERE ref_id = 2;
UPDATE referenced_table SET id = 15 WHERE id = 2;
ERROR:  canceling the transaction since it was involved in a distributed deadlock
```

### 5. Self deadlocks with a single distributed command:
```SQL
ALTER TABLE referencing_table DROP COLUMN ref_id ;
ERROR:  canceling the transaction since it was involved in a distributed deadlock

TRUNCATE referenced_table CASCADE;
ERROR:  canceling the transaction since it was involved in a distributed deadlock

ALTER TABLE referencing_table DROP CONSTRAINT test_constraint;
ERROR:  canceling the transaction since it was involved in a distributed deadlock
```

### My high-level proposal to fix these transaction related issues

* Citus should be able to track which tables/placements are modified in a transaction.
* Citus should have two modes for query execution: `parallel` and `sequential`
    * We already support such a distinction for multi shard modification queries via `show citus.multi_shard_modify_mode;`
    * We could potentially deprecate this GUC and add a new one: `citus.multi_shard_query_execution_mode`
* Whenever we detect that we hit 1, 2 or 3, we switch the mode to `sequential` and for all new commands to use the already opened connections.
* Implementing  4 is not possible if the first query already does a parallel query. We could only error out saying that set the mode to sequential and try again.
* Implementing 5 is just setting the mode to `sequential` when we realize that the DDL operation on a distributed table that is refers a reference table and the DDL is related with the fkey columns.
* I think fixing the issues related to `COPY` and `coordinator insert select` is not easily possible because `COPY` always opens a connection per placement. Changing that is beyond the scope/estimate of this task, I believe.
* Implementing this in real-time executor will probably requires some more thinking, but, seems like we should simply change the connection flag to `SESSION_LIFESPAN` when needed.
* I don't see any reason to acquire shard resource locks for this task. Note that we don't allow hash distributed tables with rep > 1 or reference table to reference table fkey with this task.


### The other refactoring/testing required to allow foreign keys to reference tables as follows:

- [ ] Refactor `ErrorIfUnsupportedForeignConstraint()` to allow reference tables be in the referenced table of a foreign key
    - [ ] Allow non-distribution keys of hash distributed tables to refer any column on the reference tables
    - [ ] Allow `ON UPDATE/DELETE SET ..` clauses if the foreign key doesn't include distribution key on the hash distributed tables
- [ ] Write a function that creates the graph of referential integrity constraints. Because, it is not enough to trace only the direct foreign key references. Instead, we should be able to track transitive relations of foreign keys. For example, if a reference table is referred by a distributed table, which is also referred by another distributed table, we should be applying the similar restriction for the second distributed table as well.  
    - [ ] Given a relation, check whether it has (transitive) foreign keys to another relation
    - [ ] Given a relation, check whether another table (transitively) refers to it
- [ ] Make sure that the code paths that rely on foreign key creation among co-located tables are refactored to consider reference tables:
    - [ ] InterShardDDLTaskList()
    - [ ] copyShardForeignConstraintCommandList() -- rebalancer and tenant isolation
    - [ ] Make sure that rebalancer and tenant isolation works fine
- [ ] Make sure that MX works fine
- [ ] Make sure that partitioned tables works fine
- [ ] Make sure that we can create reference table and distributed tables on different schemas
- [ ] Make sure that multi column foreign keys works expected
- [ ] Make sure that both `ALTER TABLE` and `CREATE TABLE` successfully creates the foreign keys on the placement

**Comment by onderkalaci** [comment]
The following is a pseduo-code to implement the transactional restrictions that'd arise due to foreign keys to reference tables. It is likely to change as we iterate on the design or cover some other edge cases. However, my basic proposal is to call `CoordinatedTransactionCheckAccess()` per `relationAccess` in the execution. Do not keep the previous state, instead simply use `pg_locks` as below.

```python
class RelationAccess
	Oid relationId
	int AccessType
	bool multiShard


##
## rollsback the transaction or swith the mode to sequential execution
## should be called before the executor starts (e.g., CitusModifyBeginScan and CitusSelectBeginScan)
##
def CoordinatedTransactionCheckAccess(relationAccess myRelation)
	
	# we could use pg_locks to gather the previous table accesses
	# so don't need to keep track of them seperately
	relationsModifiedInTheTx =  select relation::regclass, mode from pg_locks where pid = pg_backend_pid() and granted and locktype = 'relation' and mode >= 'RowExclusiveLock'


	if myRelation is referenceRelation:
		hashRelationsThatMyRelationCascadesTo = RelationsThatMyRelationCascades(relationAccess.relationId).filter(hashRelation)
	
		if AccessType == UPDATE_DELETE_DDL AND relationsModifiedInTheTx.intersects(hashRelationsThatMyRelationCascadesTo):
			error("cannot mofify the reference since other hash distributed relations that the reference table cascades already modified")

	else if myRelation is a hashRelation:

		# we're only interested in multi shard commands
		if !myRelation.multiShard 
			return

		referenceRelationsThatCascadesToMyRelation = RelationsThatCascadesToMyRelation(relationAccess.relationId).filter(referenceRelation)

		if relationsModifiedInTheTx.intersects(referenceRelationsThatCascadesToMyRelation):
			if relationAccess.AccessType == COPY:
				error("cannot copy after modification to a reference table that cascades to this relation")
			else:
				# we could only use the same connections to execute the query
				citus.multi_shard_query_execution_mode = 'sequential'
```

I'm putting all these so that we could have a design review by @marcocitus on these notes.

**Comment by furkansahin** [comment]
I am writing this comment to just track the cases that I worked on about foreign key creation
- [ ] `alter table t2 add constraint t2_j_fkey FOREIGN KEY (j) REFERENCES t1(j) ON DELETE CASCADE;` should create foreign keys in all of the shards.
- [ ] ```select create_re

**Comment by mkurz** [comment]
Fixed by #2692?

**Comment by onderkalaci** [comment]
> Fixed by #2692?

No, it wasn't. To clarify, Citus [supports](http://docs.citusdata.com/en/v8.1/develop/reference_ddl.html#adding-removing-constraints) foreign keys between distributed tables and foreign keys from distributed tables to reference tables. We keep this issue open because Citus currently does not support foreign keys between reference tables. We'll close the issue once we implement it.

#2692 in short is that we're implementing a new executor for Citus (#2435), and 2692 expands foreign key support to that executor. If you look closely, the mentioned PR is merged into `unified_executor` branch, not master.

**Comment by mkurz** [comment]
@onderkalaci 
>  We keep this issue open because Citus currently does not support foreign keys between reference tables. We'll close the issue once we implement it.

#2864 was [merged](https://github.com/citusdata/citus/commit/a5b087c89bc9a601dd6462a9348ef4c84e990921). I guess this can be closed now?

**Comment by onderkalaci** [comment]
> #2864 was merged. I guess this can be closed now?

I think so! After almost 2 years, we seem to be good to close this issue.
